File: 6549.cpp
#include<iostream>
#include<cstdio>
#include<algorithm>

using namespace std;
typedef long long lld;

lld N, h[101010];

lld findIntersection(lld start, lld mid, lld end){
    lld i=mid, j=mid+1, minHeight=min(h[i], h[j]), width=2, maxArea=width*minHeight;
    

    while(i>start && j<end){
        if(h[i-1] < h[j+1]) minHeight = min(minHeight, h[++j]);
        else minHeight = min(minHeight, h[--i]);
        maxArea = max(maxArea, ++width*minHeight);
    }
    while(i>start){
        minHeight = min(minHeight, h[--i]);
        maxArea = max(maxArea, ++width*minHeight);
    }
    while(j<end){
        minHeight = min(minHeight, h[++j]);
        maxArea = max(maxArea, ++width*minHeight);
    }
    
    return maxArea;
}

lld solve(lld start, lld end){
    if(start == end) return h[start];

    lld mid = (start+end)/2;
    lld area1 = solve(start, mid);
    lld area2 = solve(mid+1, end);
    lld area3 = findIntersection(start, mid, end);

    return max(area1, max(area2, area3));
}

int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio();

    while(true){
        cin >> N;
        if(N==0) break;
        for(int i=0; i<N; i++){
            cin >> h[i];
        }
        cout << solve(0, N-1) << '\n';
    }   
}

File: 6549.py

 
def getMaxWithMiddle(A, p, q):
    m = (p+q)//2
    
    i = m
    j = m
    min_height = A[m]
    max_area = 0
    
    while True:
        min_height = min(min(A[i], A[j]), min_height)

        while (i-1) >= p and min_height <= A[i-1]:
            i -= 1
        
        while (j+1) <= q and min_height <= A[j+1]:
            j += 1

        max_area = max(((j-i+1)*min_height), max_area)

        if i > p and j < q:
            if A[i-1] <= A[j+1]:
                j += 1
            else:
                i -= 1
        elif i > p:
            i -= 1
        elif j < q:
            j += 1
        else:
            break

    return max_area

def getMaxArea(A, p, q):
    if p == q: 
        return A[p]
    
    m = (p+q)//2
    
    leftMaxArea = getMaxArea(A, p, m)
    rightMaxArea = getMaxArea(A, m+1, q)
    midMaxArea = getMaxWithMiddle(A, p, q)

    return max(max(leftMaxArea, rightMaxArea), midMaxArea)

while True:
    a = list(map(int, input().split(" ")))
    length = a[0]
    a = a[1:]

    if not length: 
        break
    
    print(getMaxArea(a, 0, length-1))